src/code/z_room.c

```diff
@@ -1,94 +1,578 @@
+    RoomShapeNormal* roomShape;
+    RoomShapeDListsEntry* entry;
     s32 i;
-    RoomMeshType0Params* meshParams;
-    GraphicsContext* gfxCtx;
-    UNK_TYPE4 pad;
-
-    gfxCtx = play->state.gfxCtx;
-    if (flags & 1) {
-        func_800BCBF4(&D_801C1D10, play);
-        gSPSegment(gfxCtx->polyOpa.p++, 0x03, room->segment);
+
+    OPEN_DISPS(play->state.gfxCtx);
+
+    if (flags & ROOM_DRAW_OPA) {
+        func_800BCBF4(&sZeroVec, play);
+        gSPSegment(POLY_OPA_DISP++, 0x03, room->segment);
         func_8012C268(play);
-        gSPMatrix(gfxCtx->polyOpa.p++, &gIdentityMtx, G_MTX_MODELVIEW | G_MTX_LOAD);
+        gSPMatrix(POLY_OPA_DISP++, &gIdentityMtx, G_MTX_MODELVIEW | G_MTX_LOAD);
     }
 
-    if (flags & 2) {
-        func_800BCC68(&D_801C1D10, play);
-        gSPSegment(gfxCtx->polyXlu.p++, 0x03, room->segment);
+    if (flags & ROOM_DRAW_XLU) {
+        func_800BCC68(&sZeroVec, play);
+        gSPSegment(POLY_XLU_DISP++, 0x03, room->segment);
         func_8012C2DC(play->state.gfxCtx);
-        gSPMatrix(gfxCtx->polyXlu.p++, &gIdentityMtx, G_MTX_MODELVIEW | G_MTX_LOAD);
+        gSPMatrix(POLY_XLU_DISP++, &gIdentityMtx, G_MTX_MODELVIEW | G_MTX_LOAD);
     }
 
-    mesh = &room->mesh->type0;
-    meshParams = Lib_SegmentedToVirtual(mesh->paramsStart);
-    for (i = 0; i < mesh->count; i++) {
-        if ((flags & 1) && (meshParams->opaqueDl != NULL)) {
-            gSPDisplayList(gfxCtx->polyOpa.p++, meshParams->opaqueDl);
+    roomShape = &room->roomShape->normal;
+    entry = Lib_SegmentedToVirtual(roomShape->entries);
+    for (i = 0; i < roomShape->numEntries; i++) {
+        if ((flags & ROOM_DRAW_OPA) && (entry->opa != NULL)) {
+            gSPDisplayList(POLY_OPA_DISP++, entry->opa);
         }
 
-        if ((flags & 2) && (meshParams->translucentDl != NULL)) {
-            gSPDisplayList(gfxCtx->polyXlu.p++, meshParams->translucentDl);
+        if ((flags & ROOM_DRAW_XLU) && (entry->xlu != NULL)) {
+            gSPDisplayList(POLY_XLU_DISP++, entry->xlu);
         }
 
-        meshParams++;
+        entry++;
     }
+
+    CLOSE_DISPS(play->state.gfxCtx);
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_room/Room_DrawType2Mesh.s")
+typedef enum {
+    /* 0 */ ROOM_CULL_DEBUG_MODE_OFF,
+    /* 1 */ ROOM_CULL_DEBUG_MODE_UP_TO_TARGET,
+    /* 2 */ ROOM_CULL_DEBUG_MODE_ONLY_TARGET
+} RoomCullableDebugMode;
+
+typedef struct RoomShapeCullableEntryLinked {
+    /* 0x00 */ RoomShapeCullableEntry* entry;
+    /* 0x04 */ f32 boundsNearZ;
+    /* 0x08 */ struct RoomShapeCullableEntryLinked* prev;
+    /* 0x0C */ struct RoomShapeCullableEntryLinked* next;
+} RoomShapeCullableEntryLinked; // size = 0x10
+
+// TODO: 127 is an arbitrarily chosen number to make the stack sorta work
+#define ROOM_SHAPE_CULLABLE_MAX_ENTRIES 127
+
+#ifdef NON_MATCHING
+// Small regalloc, likely related to temp usage and scoping
+void Room_DrawCullable(PlayState* play, Room* room, u32 flags) {
```

Comment:
I don't suppose this yields to any do while (0) shenanigans in the macros?

Reply:
Nothing stands out, but I can try looking at it. I don't think I looked at debug for this either, pretty sure we can get `ROOM_SHAPE_CULLABLE_MAX_ENTRIES` from it most likely, add it to the TODO